File: src/main/java/org/kite9/diagram/visualization/compaction/rect/HierarchicalCompactionStep.java
package org.kite9.diagram.visualization.compaction.rect;

import org.kite9.diagram.visualization.compaction.AbstractCompactionStep;
import org.kite9.diagram.visualization.compaction.Compaction;
import org.kite9.diagram.visualization.compaction.Compactor;
import org.kite9.diagram.visualization.compaction.Embedding;
import org.kite9.diagram.visualization.display.CompleteDisplayer;

/**
 * This makes sure that compaction proceeds bottom-up through the diagram.
 * 
 * @author robmoffat
 *
 */
public class HierarchicalCompactionStep extends AbstractCompactionStep {

	public HierarchicalCompactionStep(CompleteDisplayer cd) {
		super(cd);
	}

	@Override
	public void compact(Compaction c, Embedding e, Compactor rc) {
		for (Embedding e2: e.getInnerEmbeddings()) {
			log.send("Compacting: "+e2);
			rc.compact(e2, c);
		}
	}
	
	
	
	@Override
	public String getPrefix() {
		return "HCS ";
	}

}

File: k9layout/src/main/kotlin/org/kite9/diagram/visualization/compaction/rect/HierarchicalCompactionStep.kt
package org.kite9.diagram.visualization.compaction.rect

import org.kite9.diagram.visualization.compaction.AbstractCompactionStep
import org.kite9.diagram.visualization.compaction.Compaction
import org.kite9.diagram.visualization.compaction.Compactor
import org.kite9.diagram.visualization.compaction.Embedding
import org.kite9.diagram.visualization.display.CompleteDisplayer

/**
 * This makes sure that compaction proceeds bottom-up through the diagram.
 *
 * @author robmoffat
 */
class HierarchicalCompactionStep(cd: CompleteDisplayer) : AbstractCompactionStep(cd) {

    override fun compact(c: Compaction, e: Embedding, rc: Compactor) {
        log.send("Compacting: $e")
        for (e2 in e.innerEmbeddings) {
            rc.compact(e2, c)
        }
    }

    override val prefix: String
        get() = "HCS "
}